File: Dynamic Algorithm/python/gridTraveler.py
import time
start_time = time.time()

memo={}

def gridTraveler(n,m):
    if (n==1 and m==1): return 1
    if n==0 or m==0: return 0

    key=str(n)+','+str(m)
    if key in memo:
        return memo[key]

    t=gridTraveler(n-1,m)+gridTraveler(n,m-1)
    memo[key]=t

    return t

def gridTraveler2(n,m):
    if (n==1 and m==1): return 1
    if n==0 or m==0: return 0

    key1=str(n)+','+str(m)
    key2=str(m)+','+str(n)
    
    if key1 in memo:
        return memo[key1]
    if key2 in memo:
        return memo[key2]

    t=gridTraveler2(n-1,m)+gridTraveler2(n,m-1)
    if n<m:
        memo[key1]=t
    else: memo[key2]=t
    # print(memo)
    return t

print(gridTraveler2(250,100))
print(len(memo))
print(time.time() - start_time)

File: Dynamic Algorithm/c++/gridTraveler.cpp
#include <bits/stdc++.h>
#define ll long double
using namespace std;

struct dict
{
    int key1;
    int key2;
    ll val;
};

vector<dict>memo;

int findKey(int k1, int k2)
{
    if (k1>k2)
        swap(k1,k2);
    int n=memo.size();
    for (int i=0;i<n;i++)
        if (k1==memo[i].key1 && k2==memo[i].key2)
            return i;
    return -1;
}

ll gridTraveler(int m, int n)
{
    if (m==1&&n==1) return 1;
    if (m==0||n==0) return 0;

    int p=findKey(m,n);
    if (p!=-1)
        return memo[p].val;

    ll t=gridTraveler(m-1,n)+gridTraveler(m,n-1);

    if(m<n)
        memo.push_back({m,n,t});
    else memo.push_back({n,m,t});

    return t;
}

ll gridTraveler_ver2(int m, int n)
{
    ll memo2[m+1][n+1];
    for (int i=0;i<=m;i++)
        memo2[i][0]=0;
    for (int j=0;j<=n;j++)
        memo2[0][j]=0;
    memo2[1][1]=1;

    for (int i=1,j=2;i<=m;i++,j=1)
        for (;j<=n;j++)
            memo2[i][j]=memo2[i-1][j]+memo2[i][j-1];

//    cout<<endl;
//    for (int i=0;i<=m;i++)
//    {
//        for (int j=0;j<=n;j++)
//            printf("%.0Lf ",memo2[i][j]);
//        cout<<endl;
//    }

    return memo2[m][n];
}

int main ()
{
    int a=250,b=100;
    gridTraveler_ver2(a,b);

    //Ver2 is better than Ver1
    printf("%.0Lf - %.0Lf | memo-size = ",gridTraveler(a,b),gridTraveler_ver2(a,b));
    cout<<memo.size();
}

//https://www.youtube.com/watch?v=oBt53YbR9Kk&list=PL5P21V35wKDkzAOVpZ6qfqrVq_43yfc-J&index=1&t=2319s
